runtime-sdk/src/modules/rofl/policy.rs

```diff
@@ -52,3 +129,190 @@ pub enum FeePolicy {
     #[default]
     EndorsingNodePays = 2,
 }
+
+/// An evaluator of an endorsement policy which decides whether the node that endorsed the given
+/// enclave is acceptable accoording to a policy.
+pub trait EndorsementPolicyEvaluator {
+    /// Verify the given endorsed TEE capability against an endorsement policy.
+    fn verify<C: Context>(
+        ctx: &C,
+        policy: &[Box<AllowedEndorsement>],
+        ect: &EndorsedCapabilityTEE,
+        metadata: &BTreeMap<String, String>,
+    ) -> Result<Option<Node>, Error> {
+        let endorsing_node_id = ect.node_endorsement.public_key;
+
+        // Attempt to resolve the node that endorsed the enclave. It may be that the node is not
+        // even registered in the consensus layer which may be acceptable for some policies.
+        let endorsing_node = || -> Result<Option<Node>, Error> {
+            let registry = RegistryImmutableState::new(ctx.consensus_state());
+            let node = registry
+                .node(&endorsing_node_id)
+                .map_err(|_| Error::UnknownNode)?;
+            let node = if let Some(node) = node {
+                node
+            } else {
+                return Ok(None);
+            };
+            // Ensure node is not expired.
+            if node.expiration < ctx.epoch() {
+                return Ok(None);
+            }
+
+            Ok(Some(node))
+        }()?;
+
+        Self::verify_atoms(
+            ctx,
+            EndorsementPolicyOperator::Or,
+            policy,
+            ect,
+            endorsing_node_id,
+            &endorsing_node,
+            metadata,
+            4,
+        )?;
+
+        Ok(endorsing_node)
+    }
+
+    /// Verify multiple endorsement policy atoms.
+    #[allow(clippy::too_many_arguments)]
+    fn verify_atoms<C: Context>(
+        ctx: &C,
+        op: EndorsementPolicyOperator,
+        policy: &[Box<AllowedEndorsement>],
+        ect: &EndorsedCapabilityTEE,
+        endorsing_node_id: PublicKey,
+        endorsing_node: &Option<Node>,
+        metadata: &BTreeMap<String, String>,
+        max_depth: usize,
```

Comment:
We already count atoms, so why is this useful? Because of `And` and `Or` I would also check width (e.g. use normal counter).

Reply:
Why is what useful, depth? Defense in depth against too deep stacks. If atom count is 32 you could create policies which go 32 levels deep which I don't think is desired.